master/internal/api_workspace.go

```diff
@@ -519,6 +818,266 @@ func (a *apiServer) PatchWorkspace(
+	return namespaceMeta
+}
+
+func (a *apiServer) SetWorkspaceNamespaceBindings(ctx context.Context,
+	req *apiv1.SetWorkspaceNamespaceBindingsRequest,
+) (*apiv1.SetWorkspaceNamespaceBindingsResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := a.GetWorkspaceByID(ctx, req.WorkspaceId, *curUser, false)
+	if err != nil {
+		return nil, err
+	}
+
+	var finalRes *apiv1.SetWorkspaceNamespaceBindingsResponse
+	err = db.Bun().RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+		res, err := a.setWorkspaceNamespaceBindings(ctx, req, &tx, curUser, w)
+		if err != nil {
+			return err
+		}
+		finalRes = res
+		return nil
+	})
+	return finalRes, err
+}
+
+func (a *apiServer) setResourceQuotas(ctx context.Context,
+	req *apiv1.SetResourceQuotasRequest, tx *bun.Tx, curUser *model.User,
+	w *workspacev1.Workspace,
+) (*apiv1.SetResourceQuotasResponse, error) {
+	license.RequireLicense("set resource quota")
+	if err := workspace.AuthZProvider.Get().
+		CanSetResourceQuotas(ctx, *curUser, w); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	autoGeneratedNamespace, err := getAutoGeneratedNamespaceName(ctx, int(w.Id), tx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting generated namespace name for workspace %s: %w",
+			w.Name, err)
+	}
+
+	for clusterName, quota := range req.ClusterQuotaPairs {
+		var wsns model.WorkspaceNamespace
+		err = tx.NewSelect().
+			Model(&model.WorkspaceNamespace{}).
+			Where("workspace_id = ?", req.Id).
+			Where("cluster_name = ?", clusterName).
+			Scan(ctx, &wsns)
+		if err != nil {
+			if err == sql.ErrNoRows {
+				defaultNamespace, err := a.m.rm.DefaultNamespace(clusterName)
+				if err != nil {
+					return nil, err
+				}
+				if defaultNamespace != autoGeneratedNamespace {
+					return nil, errors.New("cannot set resource quota on a namespace that was " +
+						"not auto-created by Determined.")
+				}
+			}
+			return nil, fmt.Errorf("error getting workspace-namespace binding: %w", err)
+		}
+
+		if wsns.Namespace != *autoGeneratedNamespace {
+			return nil, errors.New("cannot set quota on a workspace that is not bound to an " +
+				"auto-created namespace")
+		}
+		err = a.m.rm.SetResourceQuota(int(quota), wsns.Namespace, clusterName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create quota in Kubernetes: %w", err)
+		}
+	}
+
+	return &apiv1.SetResourceQuotasResponse{}, nil
+}
+
+func (a *apiServer) SetResourceQuotas(ctx context.Context,
+	req *apiv1.SetResourceQuotasRequest,
+) (resp *apiv1.SetResourceQuotasResponse, err error) {
+	defer func() {
+		if recover() != nil {
+			err = status.Error(codes.InvalidArgument, "Setting the resource quota on a workspace "+
+				"is an Enterprise-Edition feature")
+		}
+	}()
+	license.RequireLicense("set resource quota")
+
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tx, err := db.Bun().BeginTx(ctx, nil)
```

Comment:
This transaction is never committed and never rolled back. For safety, I recommend doing something like
```go
tx, err := db.Bun().BeginTx(ctx, nil)
if err != nil {
	return nil, err
}
defer tx.Rollback()

// use the tx for useful stuff

err = tx.Commit()
if err != nil {
	logrus.WithError(err).Error("error committing transaction while setting resource quotas")
	return status.Error(codes.Internal, "internal database error")
}

return nil
```